### DISCUSSION TITLE
What command was sent to px4

### DISCUSSION THREAD

**Discussion by sytnocui** [aerorobotics/neural-fly#5] [Q&A]
Hello! May I ask what control instructions are you sending to PX4? In the paper, you pointed out that the original control code of PX4 was used without modification. However, in the control algorithm section of the paper (page 20, and Figure 2), you provided a control formula based on the combination of model feedforward control and PID feedback control (formula 7), and the calculated control output are the control force and control torque in generalized coordinates. But as far as I know, the commands sent by Mavros to px4 are usually the expected position, expected speed, or expected attitude. May I ask what commands are sent to px4 during the algorithm deployment phase and do you have any additional processing?

Furthermore, as I understand it,  the PID algorithm is run in Raspberry Pi and px4 respectively, while the PID mentioned in the paper refers to the one running in Raspberry Pi. Is this my understanding correct?

As you see，my English is not very well.If you know Chinese, the following words may better understand my question. Thank you very much.

您好！请问你们发送给PX4的是什么控制指令呢？在论文中，你们指出了使用的是PX4的未经修改的原始控制代码。但是在论文的控制算法章节（第20页，和图2），你们给出了基于模型前馈控制与PID反馈控制结合的控制公式（公式7），计算得到的控制量是广义坐标的控制力与控制力矩。但据我所知，mavros给px4发送的指令一般是期望位置，期望速度，或者期望姿态，请问你们控制算法部署阶段给px4发送的是什么指令，有没有额外的处理？
另外，我现在理解的是在树莓派中和px4中分别运行各自的pid反馈控制，而论文中指出的与模型前馈结合的PID控制指的是在树莓派中运行的那一份，请问我这个理解对吗？

**Comment by mtoc12** [comment]
Hi @sytnocui, thanks for the question. Yes, we send quaternion and thrust commands. See the end of the Materials and Methods, Implementation Details, Quadrotor Dynamics:

"Note that the quadrotor attitude dynamics ([Eq. 13b](https://www.science.org/stoken/author-tokens/ST-463/full#E13b)) is also a special case of [Eq. 1](https://www.science.org/stoken/author-tokens/ST-463/full#E1) ([15](https://www.science.org/stoken/author-tokens/ST-463/full#core-R15), [51](https://www.science.org/stoken/author-tokens/ST-463/full#core-R51)), and thus our method can be extended to attitude control. We implement our method in the position control loop; that is, we use our method to compute a desired force ud. Then, the desired force is decomposed into the desired attitude Rd and the desired thrust Td using kinematics (see [Fig. 2](https://www.science.org/stoken/author-tokens/ST-463/full#F2)). Then, the desired attitude and thrust are sent to the onboard PX4 flight controller."

We removed most of the PID results because standard practice would be to use the onboard PID controller for tracking position. The PID results shown are for a PID tracking position, sending target thrust and quaternion commands to the PX4 in OFFBOARD_MODE. This matches the implementation of our method, so it is a good comparison of the method, and still remains comparable to the state of practice.

**Reply by sytnocui** [reply]
Thanks for your reply. Finally, I want to confirm a point.
I note the statement in your reply:
"Then, the desired force is decomposed into the desired attitude Rd and the desired thrust Td using kinematics (see [Fig. 2]). Then, the desired attitude and thrust are sent to the onboard PX4 flight controller."

Is that mean the desired thrust Td is completely not used?
